daceml/pytorch/module_codegen.py

```diff
@@ -0,0 +1,287 @@
+    ptr_init_code += "\n    // setup constant arguments\n"
+
+    # initialize all remaining parameters
+    remaining = set(arglist).difference(
+        itertools.chain(input_names, output_names))
+    for name in remaining:
+
+        # remaining args must be constants
+        if name not in module.dace_model.clean_weights:
+            raise ValueError(
+                f"Cannot generate PyTorch module C++ code: SDFG argument {name} is not an input or output"
+                f" of the PyTorch Module, and not a constant.")
+        if arglist[name].total_size > 1000:
+            raise ValueError(
+                f"Cannot generate PyTorch module C++ code: SDFG argument {name} is not an input or output"
+                f" of the PyTorch Module, and is too large.")
+
+        value = module.dace_model.clean_weights[name]
+        ptr_init_code += f"    {constant_initializer_code(name + '_ptr', arglist[name], value)}\n"
+
+    arguments = ", ".join(f"{n}_ptr" for n in arglist)
+    init_arguments = ", ".join(f"{n}_ptr" for n, desc in arglist.items()
+                               if isinstance(desc, data.Scalar))
+
+    return ptr_init_code, arguments, init_arguments
+
+
+def constant_initializer_code(name: str, desc: data.Data, value) -> str:
+    if isinstance(desc, data.Array):
+        iterator = np.nditer(value.cpu().numpy(), order="C")
+        return f"{desc.dtype.ctype} {name}[{desc.total_size}] = {{{', '.join(str(e) for e in iterator)}}};"
+    else:
+        return f"{desc.dtype.ctype} {name} = {str(value.item())};"
+
+
+def code_for_backward_function(module: 'daceml.pytorch.DaceModule',
+                               forward_sdfg: dace.SDFG,
+                               backward_sdfg: dace.SDFG) -> str:
+    pass
+
+
+def code_for_module(module: 'daceml.pytorch.DaceModule') -> str:
+    """ Generate the code for an operator that calls the sdfgs in the module.
+
+        :param module: the module
+    """
+
+    inputs, outputs = get_arglist(module)
+    sdfg_name = module.sdfg.name
+
+    if module.backward:
+        raise NotImplemented("todo")
+    else:
+        ptr_init_code, sdfg_call_arguments, init_arguments = argument_codegen(
+            module, inputs, outputs)
+        return f"""
+#include <torch/torch.h>
+#include <torch/script.h>
+#include "{sdfg_name}.h"
+
+using torch::Tensor;
+using torch::DeviceType;
+using torch::autograd::tensor_list;
+using torch::autograd::AutogradContext;
+
+TORCH_LIBRARY(daceml_{sdfg_name}, m) {{
+    m.def("{sdfg_name}(int handle_ptr, {", ".join('Tensor ' + arg for arg in inputs)}) -> {'Tensor' if len(outputs) == 1
+        else "(" + ", ".join(['Tensor'] * len(outputs)) + ")"}");
+}}
+
+// function definition
+{"Tensor" if len(outputs) == 1 else f"std::tuple<{', '.join(['Tensor'] * len(outputs))}>"}
+{sdfg_name}(int64_t handle_ptr, {",".join(f"const Tensor& {name}_" for name in inputs)}) {{
+
+    // TODO check contiguous
```

Comment:
important, but can be split into two: inputs can be checked for `is_contiguous()` in C++, and parameters are done on the python side prior, to get strides?

Reply:
good point